### DISCUSSION TITLE
Artemis v2.x defining topics and queues

### DISCUSSION THREAD

**Discussion by Renboo** [jmstoolbox/jmstoolbox#181] [Ideas]
Hello!

Let me tell about ActiveMQArtemis2QManager.class. I found algorithm of defining topic and queues in it. It is usually should works well, but I wouldn't be myself if I didn't work on exceptional situations and notes :smile: 

Now all my queues in Artemis MQ defined as topics. Thats why because delivery mode of them are arrays: 
`["MULTICAST", "ANYCAST"]`
and queues have arrays of 1 element with same name.
Of cause you could tell "THIS IS ANTIPATTERN!" and you will be right. But my server is not configurable by me and i could not fix that. I have to live with this.

Some useful pages to read from official documentation:
1. [Apache ActiveMQ Artemis Addressing and Queues](https://activemq.apache.org/components/artemis/documentation/2.0.0/address-model.html)
2. [Mapping JMS Concepts to the Core API](https://activemq.apache.org/components/artemis/documentation/2.0.0/jms-core-mapping.html)

Main idea is note:

> It is possible to define more than one routing type per address, but this typically results in an anti-pattern and is therefore not recommended. If an address does use both routing types, however, and the client does not show a preference for either one, the broker typically defaults to the anycast routing type. The one exception is when the client uses the MQTT protocol. In that case, the default routing type is multicast.

That situation forced me to turn to the documentation and help to fix your algorithm of defining topic and queues.
I will take timeout to think. And will try to return with new algorithm.
I would be grateful if you have any more useful documentation on this issue.

**Comment by Renboo** [comment]
First idea is update sendAdminMessage method to:


````
@SuppressWarnings("unchecked")
private <T> T sendAdminMessage(Class<T> clazz,
		Session sessionJMS,
		QueueRequestor requestorJMS,
		String resourceName,
		String methodName) throws Exception {
  Message m = sessionJMS.createMessage();
  JMSManagementHelper.putAttribute(m, resourceName, methodName);
  Message r = requestorJMS.request(m);
  if (String[].class == clazz) {
	 String strResult = (String) JMSManagementHelper.getResult(r);
	 return (T) strResult.replace("[\"", "").replace("\"]", "").replace("\",\"", "").split(",");
  }
  return (T) JMSManagementHelper.getResult(r);
}
````

**Comment by Renboo** [comment]
I'm not entirely sure because could not test more variants now but it works for me:

```
         List<String> deliveryMode = List.of(sendAdminMessage(String[].class,
                                                sessionJMS,
                                                requestorJMS,
                                                ResourceNames.ADDRESS + addressName,
                                                getRoutingTypeMtd));

         Object[] queues = sendAdminMessage(Object[].class,
                                            sessionJMS,
                                            requestorJMS,
                                            ResourceNames.ADDRESS + addressName,
                                            "queueNames");

         log.debug("addressName: {} deliveryMode: {} queues: {}", addressName, deliveryMode, queues);

         if (deliveryMode.size() == 1) {

            // MULTICAST addresses are Topics with with zero or more queues
            if (deliveryMode.contains("MULTICAST")) {
               log.debug("addressName: {} is a Topic", addressName);
               listTopicData.add(new TopicData((String) addressName));
               continue; // DF not sure of this..
            }

            // ANYCAST addresses are Queues with address with name=(the JMS queue name)
            if (deliveryMode.contains("ANYCAST") && (queues.length == 1) && (queues[0].equals(addressName))) {
               log.debug("addressName: {} is a Queue", addressName);
               listQueueData.add(new QueueData((String) addressName));
               continue;
            }

         }

         // Anipattern but broker typically defaults to ANYCAST what means Queue
         if (deliveryMode.size() > 1) {
            log.debug("addressName: {} is a Queue", addressName);
            listQueueData.add(new QueueData((String) addressName));
            continue;
         }

         // ANYCAST addresses with no queues are ... (I don't know, ignore them)
         if (queues.length == 0) {
            log.warn("addressName: {} is ANYCAST with no queues, Ignore it.", addressName);
            continue;
         }

         // Other ANYCAST adresses are Topics
         log.debug("addressName: {} is a Topic (ANYCAST with Queues that do not match address name", addressName);
         listTopicData.add(new TopicData((String) addressName));
```

Please make more tests if you could.

**Comment by titou10titou10** [comment]
I don't really understand how your topics/queues are defined.
So your queues are defined with deliveryMode = ["MULTICAST", "ANYCAST"]. right?
How are defined your topics?

To summarize, can you tell how are mapped the following combination for you?

    ["MULTICAST", "ANYCAST"]                : queue or topic?
    ["MULTICAST"]                           : queue or topic?
    ["ANYCAST"] with no queue name          : queue or topic?
    ["ANYCAST"] with 1 queue name           : queue or topic?
    ["ANYCAST"] with more than 1 queue name : queue or topic?

**Comment by Renboo** [comment]
I think that correct variant is:

```
["MULTICAST", "ANYCAST"]                : queue
["MULTICAST"]                           : topic
["ANYCAST"] with 1 queue with same name : queue
["ANYCAST"] with more than 1 queue name : queue
["ANYCAST"] with no queue name          : queue or topic? is it possible? may it be ignored? 
```

**Comment by titou10titou10** [comment]
I commit some code to reflect that. you can download the beta version here:
https://github.com/jmstoolbox/jmstoolbox/actions/runs/10912652882
https://github.com/jmstoolbox/jmstoolbox/actions/runs/10912652882/artifacts/1945323394

**Comment by Renboo** [comment]
Thank you for update.

I have tested you code. It main cases it works ok, but i have lost 1 queue.
Information about it:

`addressName: 09d86191-3698-4388-8bd0-a7590708b77c deliveryMode: [ANYCAST] queues: [09d86191-3698-4388-8bd0-a7590708b77c]`

It should be selected by that condition:
`if ((queues.length == 1) && (queues[0].equals(addressName)))`

Do you have any ideas?

**Comment by titou10titou10** [comment]
Is this queue marked as temporary and in the preference you said to not show system/temporary objects?
Show all the logs about the processing of this Queue by JTB

**Comment by Renboo** [comment]
You are right. I enabled system/temporary objects in preference and got that queue.
Now all works good as we have discussed.
Thank you for help! I will wait for new release.
